tensorflow/core/framework/dataset.h

```diff
@@ -273,6 +371,236 @@ class GraphDefBuilderWrapper {
 class StatsAggregator;
 class FunctionHandleCache;
 
+typedef std::map<string, std::vector<EparallaxTensorIndex*>*> IndexTree;
+class MultiLevelIndexTree {
+ public:
+  void Push(EparallaxTensorIndex* index) {
+    std::vector<EparallaxTensorIndex*>* indices = Get(
+        index->iterator_id(), ToString(*index->parent_indices()));
+    indices->push_back(index);
+  }
+
+  bool Contains(EparallaxTensorIndex *index) {
+    std::vector<EparallaxTensorIndex*>* indices = Get(
+        index->iterator_id(), ToString(*index->parent_indices()));
+    for (auto buffered_index : *indices) {
+      if (*index == *buffered_index) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  void ClearParent(string iterator_id) {
+    string key;
+    IndexTree* tree;
+    for (auto& it : tree_) {
+      key = it.first;
+      tree = it.second;
+      if (key.substr(0, iterator_id.length()) == iterator_id) {
+        if (key.length() > iterator_id.length()) {
+          tree->clear();
+        }
+      }
+    }
+  }
+
+  std::vector<EparallaxTensorIndex*>* Get(string iterator_id,
+                                          string parent_indices) {
+    auto it = tree_.find(iterator_id);
+    if (it == tree_.end()) {
+      IndexTree* tree = new IndexTree;
+      std::vector<EparallaxTensorIndex*>* q =
+          new std::vector<EparallaxTensorIndex*>;
+      tree->insert(std::make_pair(parent_indices, q));
+      tree_.insert(std::make_pair(iterator_id, tree));
+      return q;
+    } else {
+      IndexTree* tree = it->second;
+      if (tree->find(parent_indices) == tree->end()) {
+        std::vector<EparallaxTensorIndex*>* q =
+            new std::vector<EparallaxTensorIndex*>;
+        tree->insert(std::make_pair(parent_indices, q));
+        return q;
+      } else {
+        return tree->find(parent_indices)->second;
+      }
+    }
+  }
+
+  std::vector<std::vector<EparallaxTensorIndex*>*> GetAll() {
+    std::vector<std::vector<EparallaxTensorIndex*>*> index_vectors;
+    for (auto const& it : tree_) {
+      for (auto const& i: *it.second) {
+        index_vectors.push_back(i.second);
+      }
+    }
+    return index_vectors;
+  }
+
+ private:
+  std::map<string, IndexTree*> tree_;
+};
+
+class IndexManager {
+ public:
+  IndexManager() :
+    mu_(std::make_shared<mutex>()),
+    processed_indices_(std::make_shared<MultiLevelIndexTree>()),
+    issued_indices_(std::make_shared<MultiLevelIndexTree>()),
+    infertile_indices_(std::make_shared<MultiLevelIndexTree>()),
+    last_index_map_(
+        std::make_shared<std::map<string, EparallaxTensorIndex*>>()),
+    current_index_map_(
+        std::make_shared<std::map<string, EparallaxTensorIndex*>>()),
+    offset_map_(std::make_shared<std::map<string, int64>>()) {
+    string username(std::getenv("USER"));
```

Comment:
The checkpoint path must be received as an argument. If there are multiple checkpoints, L457~L458 would be broken.

Reply:
You have the point. I fixed this part so that we get the checkpoint directory from the environment variable `EPARALLAX_CKPT_DIR`.